frontend/src/pages/_app.tsx

```diff
@@ -6,26 +6,97 @@ import createCache from "@emotion/cache";
 import { CacheProvider } from "@emotion/react";
 import { IntlProvider } from "react-intl";
 import { useRouter } from "next/router";
-import { useMemo } from "react";
+import { useCallback, useEffect, useMemo, useState } from "react";
+import {
+  AuthenticationContext,
+  authenticationContextDefaultValue,
+  AuthenticationContextType,
+  deserializeAuthenticationContext,
+  serializeAuthenticationContext,
+} from "@/contexts/AuthenticationContext";
+import { UpdateAuthenticationContext } from "@/contexts/UpdateAuthenticationContext";
+import AuthenticationBarrier from "@/components/authentication/AuthenticationBarrier";
+
+const authenticationStateKey = "authenticationState";
 
 const cache = createCache({ key: "next" });
 
+const getInitialState = async (): Promise<AuthenticationContextType> => {
+  if (typeof localStorage === "undefined") {
+    // for SSR we return the default value
+    return authenticationContextDefaultValue;
+  }
+
+  // load the stored authentication state from localStorage
+  const storedAuthenticationState = localStorage.getItem(
+    authenticationStateKey,
+  );
+
+  return storedAuthenticationState
+    ? deserializeAuthenticationContext(
+        window.crypto.subtle,
+        JSON.parse(storedAuthenticationState),
+      )
+    : authenticationContextDefaultValue;
+};
+
 const App = ({ Component, pageProps }: AppProps) => {
-  const { locale } = useRouter();
+  const router = useRouter();
+
+  const [authenticationStateLoaded, setAuthenticationStateLoaded] =
+    useState(false);
+
+  const [authenticationState, setAuthenticationState] =
+    useState<AuthenticationContextType>(authenticationContextDefaultValue);
+
+  const updateAuthenticationState = useCallback(
+    (newState: AuthenticationContextType) =>
+      setAuthenticationState(() => {
+        // store the new state in the session storage
+        serializeAuthenticationContext(newState).then((serializedNewState) => {
+          localStorage.setItem(
+            authenticationStateKey,
+            JSON.stringify(serializedNewState),
+          );
+        });
+
+        return newState;
```

Comment:
we cannot use async await here because we need to synchronously return the new state. however, we want to update the serialized state in localstorage at the same time too. but you are right about errors!